### PR TITLE
[1692] Add endpoint to update GCSE requirements

### PR DESCRIPTION
**Pull request by davidgisbey**
### Context

Providers currently have to contact support to change GCSE requirements on their courses. 

### Changes proposed in this pull request

Adds the backend work for changing gcse requirements to the course update endpoint.

### Guidance to review

We will need to think about how we wan't to address secondary not needing science to be set and FE courses not needing GCSE requirements. Paul H said this will likely occur on the front end so i haven't added any validations on the backend for not_set or not_required etc.

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

File: spec/models/course/update_courses_gcse_requirements_spec.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    context 'with a course with updated gcse requirements' do
+      context 'when all requirements are updated' do
+        let(:params) {
+          { "english" => "expect_to_achieve_before_training_begins",
+                        "maths" => "expect_to_achieve_before_training_begins",
+                        "science" => "expect_to_achieve_before_training_begins" }
+        }
+
+        its(:english_before_type_cast) { should eq 2 }
```
Comment (benilovj) [inline spec/models/course/update_courses_gcse_requirements_spec.rb]: what's the thinking behind using the non-enum values here? IMHO it'd be easier to check that the test is doing the right thing using the enums...

File: spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  end
+  let(:organisation)      { create :organisation }
+  let(:provider)          { create :provider, organisations: [organisation], sites: [site] }
+  let(:user)              { create :user, organisations: [organisation] }
+  let(:payload)           { { email: user.email } }
+  let(:token)             { build_jwt :apiv2, payload: payload }
+  let(:course)            { create :course, provider: provider, site_statuses: [site_status], subjects: [primary_subject] }
+  let(:site_status)       { build(:site_status, :findable, site: site) }
+  let(:site)              { build(:site) }
+  let(:primary_subject)   { build(:subject, subject_name: 'primary') }
```
Comment (benilovj) [inline spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb]:
```suggestion
  let(:primary_subject)   { build(:subject, :primary) }
```

File: app/controllers/api/v2/courses_controller.rb [hunk 3]
```diff
@@ -60,8 +60,10 @@ def publishable
       def update
         update_enrichment
         update_sites
+        update_gcse_requirements
```
Comment (dankmitchell) [inline app/controllers/api/v2/courses_controller.rb]: Does this need to be a separate method? Arent we just updating the course object itself?

Comment (davidgisbey) [reply app/controllers/api/v2/courses_controller.rb]: Just discovered deserialisers 🤦‍♂. Will do this. It's way easier 👍

File: spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb (hunk 2 above)
Comment (davidgisbey) [reply spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb]: Done. Thanks

File: spec/models/course/update_courses_gcse_requirements_spec.rb (hunk 1 above)
Comment (davidgisbey) [reply spec/models/course/update_courses_gcse_requirements_spec.rb]: Have made it more explicit in the req spec

File: spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        updated_course.id = course.id
+        perform_request(updated_course)
+      end
+
+      it "returns http success" do
+        expect(response).to have_http_status(:success)
+      end
+
+      it "does not change english attribute" do
+        expect(course.reload.english).to eq(course.english)
```
Comment (timabell) [inline spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb]: isn't that always true? I think you need to keep a copy of what it was somewhere

**Comment by timabell** [comment]
Looks good to me apart from the comment on the tests

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

File: spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb (hunk 4 above)
Comment (davidgisbey) [reply spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb]: You are correct. Have addressed.

Comment (davidgisbey) [reply spec/requests/api/v2/providers/courses/update_gsce_requirements_spec.rb]: Thanks for the feedback!

**Comment by davidgisbey** [comment]
[Edit UCAS equivalency](https://trello.com/c/AKyu3oCM/1692-edit-ucas-equivalency)

**Pull request merged**